### ISSUE TITLE
Batch requests do not refresh token

### ISSUE THREAD

**Issue by 4r13** [googleapis/google-api-python-client#435]
Traceback seems very similar to https://github.com/google/google-api-python-client/issues/211, but it happens on version 1.6.3

  File "/Users/beast/Library/Python/2.7/lib/python/site-packages/cloud_utils-1.0.11-py2.7.egg/cloud_utils/list_instances.py", line 122, in gcp_filter_projects_instances
    batch.execute()
  File "/Users/beast/Library/Python/2.7/lib/python/site-packages/oauth2client/_helpers.py", line 133, in positional_wrapper
    return wrapped(*args, **kwargs)
  File "/Users/beast/Library/Python/2.7/lib/python/site-packages/googleapiclient/http.py", line 1424, in execute
    self._execute(http, self._order, self._requests)
  File "/Users/beast/Library/Python/2.7/lib/python/site-packages/googleapiclient/http.py", line 1340, in _execute
    body = self._serialize_request(request)
  File "/Users/beast/Library/Python/2.7/lib/python/site-packages/googleapiclient/http.py", line 1211, in _serialize_request
    _auth.apply_credentials(credentials, headers)
  File "/Users/beast/Library/Python/2.7/lib/python/site-packages/googleapiclient/_auth.py", line 113, in apply_credentials
    return credentials.apply(headers)
  File "/Users/beast/Library/Python/2.7/lib/python/site-packages/oauth2client/client.py", line 562, in apply
    headers['Authorization'] = 'Bearer ' + self.access_token
TypeError: cannot concatenate 'str' and 'NoneType' objects

**client version**:
Name: google-api-python-client
Version: 1.6.3
Summary: Google API Client Library for Python
Home-page: http://github.com/google/google-api-python-client/
Author: Google Inc.
Author-email: UNKNOWN
License: Apache 2.0
Location: /Users/beast/Library/Python/2.7/lib/python/site-packages
Requires: six, httplib2, uritemplate, oauth2client

**Heads up!**

We appreciate any bug reports or other contributions, but please note that this issue
tracker is only for this client library. We do not maintain the APIs that this client
library talks to. If you have an issue or questions with how to use a particular API,
you may be better off posting on Stackoverflow under the `google-api` tag.

Thank you!

**Comment by berill** [comment]
Also happening in 1.6.4.

Invoking twice the code below:
```
from oauth2client.appengine import AppAssertionCredentials
import httplib2
from apiclient import discovery
import webapp2

_OAUTH_SCOPE = 'https://www.googleapis.com/auth/webmasters.readonly'
_SITEMAP_BASE = 'YOUR_WEBSITE_HERE'

def _get_webmaster_service():
    credentials = AppAssertionCredentials(scope=_OAUTH_SCOPE)

    http = httplib2.Http()
    http = credentials.authorize(http)

    return discovery.build('webmasters', 'v3', http=http)

def do_nothing(*args, **kwargs):
	pass


class WebmasterHandler(webapp2.RequestHandler):
    def get(self, *args, **kwargs):
		webmasters_service = _get_webmaster_service()
		batch = webmasters_service.new_batch_http_request()
		for sitemap_id in [1, 2 , 3]:
			path = _SITEMAP_BASE + '/sitemap_%i.xml' % sitemap_id
			batch.add(webmasters_service.sitemaps().get(siteUrl=_SITEMAP_BASE,
	                                                    feedpath=path), callback=do_nothing)
		batch.execute()


app = webapp2.WSGIApplication([
	webapp2.Route('/' , WebmasterHandler),
	])
```

will result in:
```
INFO     2018-01-11 16:10:47,834 devappserver2.py:115] Skipping SDK update check.
INFO     2018-01-11 16:10:47,885 api_server.py:299] Starting API server at: http://localhost:64758
WARNING  2018-01-11 16:10:47,885 dispatcher.py:285] Your python27 micro version is below 2.7.12, our current production version.
INFO     2018-01-11 16:10:47,909 dispatcher.py:224] Starting module "default" running at: http://localhost:8080
INFO     2018-01-11 16:10:47,911 admin_server.py:116] Starting admin server at: http://localhost:8000
INFO     2018-01-11 16:10:54,706 discovery.py:274] URL being requested: GET https://www.googleapis.com/discovery/v1/apis/webmasters/v3/rest?userIp=%3A%3A1
INFO     2018-01-11 16:10:54,706 client.py:515] Attempting refresh to obtain initial access_token
INFO     2018-01-11 16:10:54,710 client.py:755] Refreshing access_token
INFO     2018-01-11 16:10:54,890 discovery.py:872] URL being requested: GET https://www.googleapis.com/webmasters/v3/sites/YOUR_WEBSITE_HERE/sitemaps/YOUR_WEBSITE_HERE%2Fsitemap_1.xml?alt=json
INFO     2018-01-11 16:10:54,891 discovery.py:872] URL being requested: GET https://www.googleapis.com/webmasters/v3/sites/YOUR_WEBSITE_HERE/sitemaps/YOUR_WEBSITE_HERE%2Fsitemap_2.xml?alt=json
INFO     2018-01-11 16:10:54,892 discovery.py:872] URL being requested: GET https://www.googleapis.com/webmasters/v3/sites/YOUR_WEBSITE_HERE/sitemaps/YOUR_WEBSITE_HERE%2Fsitemap_3.xml?alt=json
INFO     2018-01-11 16:10:55,000 module.py:821] default: "GET / HTTP/1.1" 200 -
INFO     2018-01-11 16:10:58,055 discovery.py:872] URL being requested: GET https://www.googleapis.com/webmasters/v3/sites/YOUR_WEBSITE_HERE/sitemaps/YOUR_WEBSITE_HERE%2Fsitemap_1.xml?alt=json
INFO     2018-01-11 16:10:58,056 discovery.py:872] URL being requested: GET https://www.googleapis.com/webmasters/v3/sites/YOUR_WEBSITE_HERE/sitemaps/YOUR_WEBSITE_HERE%2Fsitemap_2.xml?alt=json
INFO     2018-01-11 16:10:58,057 discovery.py:872] URL being requested: GET https://www.googleapis.com/webmasters/v3/sites/YOUR_WEBSITE_HERE/sitemaps/YOUR_WEBSITE_HERE%2Fsitemap_3.xml?alt=json
ERROR    2018-01-11 16:10:58,057 webapp2.py:1552] cannot concatenate 'str' and 'NoneType' objects
Traceback (most recent call last):
  File "/usr/local/Frameworks/google-cloud-sdk/platform/google_appengine/lib/webapp2-2.5.2/webapp2.py", line 1535, in __call__
    rv = self.handle_exception(request, response, e)
  File "/usr/local/Frameworks/google-cloud-sdk/platform/google_appengine/lib/webapp2-2.5.2/webapp2.py", line 1529, in __call__
    rv = self.router.dispatch(request, response)
  File "/usr/local/Frameworks/google-cloud-sdk/platform/google_appengine/lib/webapp2-2.5.2/webapp2.py", line 1278, in default_dispatcher
    return route.handler_adapter(request, response)
  File "/usr/local/Frameworks/google-cloud-sdk/platform/google_appengine/lib/webapp2-2.5.2/webapp2.py", line 1102, in __call__
    return handler.dispatch()
  File "/usr/local/Frameworks/google-cloud-sdk/platform/google_appengine/lib/webapp2-2.5.2/webapp2.py", line 572, in dispatch
    return self.handle_exception(e, self.app.debug)
  File "/usr/local/Frameworks/google-cloud-sdk/platform/google_appengine/lib/webapp2-2.5.2/webapp2.py", line 570, in dispatch
    return method(*args, **kwargs)
  File "/test/googleapiclient/test.py", line 37, in get
    batch.execute()
  File "/usr/local/Frameworks/google-cloud-sdk/platform/google_appengine/lib/oauth2client_devserver/oauth2client/util.py", line 129, in positional_wrapper
    return wrapped(*args, **kwargs)
  File "/test/googleapiclient/ext/googleapiclient/http.py", line 1424, in execute
    self._execute(http, self._order, self._requests)
  File "/test/googleapiclient/ext/googleapiclient/http.py", line 1340, in _execute
    body = self._serialize_request(request)
  File "/test/googleapiclient/ext/googleapiclient/http.py", line 1211, in _serialize_request
    _auth.apply_credentials(credentials, headers)
  File "/test/googleapiclient/ext/googleapiclient/_auth.py", line 127, in apply_credentials
    return credentials.apply(headers)
  File "/usr/local/Frameworks/google-cloud-sdk/platform/google_appengine/lib/oauth2client_devserver/oauth2client/client.py", line 576, in apply
    headers['Authorization'] = 'Bearer ' + self.access_token
TypeError: cannot concatenate 'str' and 'NoneType' objects
INFO     2018-01-11 16:10:58,065 module.py:821] default: "GET / HTTP/1.1" 500 228
```

A possible workaround is to disable cache_discovery on service discovery:
```
return discovery.build('webmasters', 'v3', http=http, cache_discovery=False)
```

**Comment by busunkim96** [comment]
@berill, you should only have to call `discovery.build()` once. Please store the service object and reuse it for all your requests. `discovery.build` is an expensive request as it fetches the whole discovery document (describing the API). 

@berill @4r13 Would either of you remember what type of credentials you were using, or anything else unique about the environment?

It looks like both of you used oauth2client.  We now recommend using [google-auth](https://github.com/googleapis/google-auth-library-python) and [google-auth-oauthlib](https://github.com/googleapis/google-auth-library-python-oauthlib) with our Python clients. Below is a minimal example showing authorization (Please reference the full documentation provided in the repositories).


```
from __future__ import print_function
import pickle
import os.path
from googleapiclient.discovery import build
from google_auth_oauthlib.flow import InstalledAppFlow
from google.auth.transport.requests import Request

# If modifying these scopes, delete the file token.pickle.
SCOPES = ['https://www.googleapis.com/auth/doubleclicksearch']

def main():
    creds = None
    # The file token.pickle stores the user's access and refresh tokens, and is
    # created automatically when the authorization flow completes for the first
    # time.
    if os.path.exists('token.pickle'):
        with open('token.pickle', 'rb') as token:
            creds = pickle.load(token)
    # If there are no (valid) credentials available, let the user log in.
    if not creds or not creds.valid:
        if creds and creds.expired and creds.refresh_token:
            creds.refresh(Request())
        else:
            flow = InstalledAppFlow.from_client_secrets_file(
                'credentials.json', SCOPES)
            creds = flow.run_local_server(port=0)
        # Save the credentials for the next run
        with open('token.pickle', 'wb') as token:
            pickle.dump(creds, token)

    service = build('doubleclicksearch', 'v2', credentials=creds)

if __name__ == '__main__':
    main()
```

**Comment by busunkim96** [comment]
Closing this issue as there has been no activity for some time. If you are experiencing this, please file a new issue. Maintainers are more likely to notice new issues than responses to closed issues.

Thanks!
